packages/api/src/bsky-agent.ts

```diff
@@ -609,3 +645,103 @@ async function updateFeedPreferences(
   })
   return res
 }
+
+/**
+ * A helper specifically for updating muted words preferences
+ */
+async function updateMutedWords(
+  agent: BskyAgent,
+  mutedWords: AppBskyActorDefs.MutedWord[],
+  action: 'upsert' | 'update' | 'remove',
+) {
+  await updatePreferences(agent, (prefs: AppBskyActorDefs.Preferences) => {
+    let mutedWordsPref = prefs.findLast(
+      (pref) =>
+        AppBskyActorDefs.isMutedWordsPref(pref) &&
+        AppBskyActorDefs.validateMutedWordsPref(pref).success,
+    )
+
+    if (mutedWordsPref && AppBskyActorDefs.isMutedWordsPref(mutedWordsPref)) {
+      if (action === 'upsert' || action === 'update') {
+        outer: for (const newItem of mutedWords) {
+          for (const existingItem of mutedWordsPref.items) {
+            if (existingItem.value === newItem.value) {
+              existingItem.targets =
+                action === 'upsert'
+                  ? Array.from(
+                      new Set([...existingItem.targets, ...newItem.targets]),
+                    )
+                  : newItem.targets
+              continue outer
+            }
+          }
+
+          // never hit `continue`, add new item
+          if (action === 'upsert') {
+            mutedWordsPref.items.push(newItem)
+          }
+        }
+      } else if (action === 'remove') {
+        let i = mutedWordsPref.items.length
+
+        while (i--) {
```

Comment:
this is a nit, but i don't love this syntax. we often use `mapDefined` for stuff like this & if the thing is going to be deleted then the callback fn just returns `undefined` 🤔

Reply:
Oh yeah I hate this too, good to know about `mapDefined` 👍